src/settings.cpp

```diff
@@ -33,48 +33,100 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include <cctype>
 #include <algorithm>
 
-Settings *g_settings = nullptr; // Populated in main()
+Settings *g_settings = nullptr;
+static SettingsHierarchy g_hierarchy;
 std::string g_settings_path;
 
-Settings *Settings::s_layers[SL_TOTAL_COUNT] = {0}; // Zeroed by compiler
 std::unordered_map<std::string, const FlagDesc *> Settings::s_flags;
 
+/* Settings hierarchy implementation */
 
-Settings *Settings::createLayer(SettingsLayer sl, const std::string &end_tag)
+SettingsHierarchy::SettingsHierarchy(Settings *fallback)
+{
+	layers.resize(1);
+	layers[0] = fallback;
+}
+
+
+Settings *SettingsHierarchy::getLayer(int layer) const
 {
-	if ((int)sl < 0 || sl >= SL_TOTAL_COUNT)
+	if (layer < 0 || layer >= layers.size())
 		throw BaseException("Invalid settings layer");
+	return layers[layer];
+}
+
 
-	Settings *&pos = s_layers[(size_t)sl];
+Settings *SettingsHierarchy::getParent(int layer) const
+{
+	assert(layer >= 0 && layer < layers.size());
+	// iterate towards the origin (0) to find the next fallback layer
+	for (int i = layer - 1; i >= 0; --i) {
+		if (layers[i])
+			return layers[i];
+	}
+
+	return nullptr;
+}
+
+
+void SettingsHierarchy::layerCreated(int layer, Settings *obj)
```

Comment:
layer should be a unsigned int

Reply:
Yes, `-1` layer values are not really necessary. `Settings::m_hierarchy` can already be used to determine whether it's part of a structure, or just any loose Settings instance in the wild.

Reply:
It does not create the settings object itself so I think `onLayerCreated` is better.